File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-8.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        chars_dict = {}
        res: int = 0
        # iterate over unique chars
        max_freq: int = 0
        left: int = 0

        # starting from right
        for right in range(len(s)):
            # if the right pointer is having the same char: count + 1
            chars_dict[s[right]] = chars_dict.get(s[right], 0) + 1
            max_freq = max(max_freq, chars_dict[s[right]])

            # we have chars in between that are not the same and greater than k to be replaced
            # invalid window
            while right-left+1 - max_freq > k:
                chars_dict[s[left]] -= 1
                left += 1
            
            res = max(res, right-left + 1)

        return res

File: Data Structures & Algorithms/spiral-matrix/submission-8.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int left = 0;
        int right = matrix[0].length - 1;

        int top = 0;
        int bottom = matrix.length - 1;

        List<Integer> res = new ArrayList(); 
        while (left <= right &&
               top <= bottom) {
            // left to right
            for (int i=left; i <= right; i++) {
                res.add(matrix[top][i]);
            }
            top++;

            // top to bottom
            for (int i=top; i <= bottom; i++) {
                res.add(matrix[i][right]);
            }  
            right--;
            if (!(left <= right && top <= bottom)) break;
            // right to left
            for (int i=right; i >= left; i--) {
                res.add(matrix[bottom][i]);
            }  
            bottom--;

            for (int i=bottom; i >= top; i--) {
                res.add(matrix[i][left]);
            }  
            left++;
               }
        return res;

    }
}
